engine/src/main/java/io/camunda/zeebe/streamprocessor/records/ModifiableRecordBatch.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.streamprocessor.records;
+
+import io.camunda.zeebe.protocol.record.RecordType;
+import io.camunda.zeebe.protocol.record.RejectionType;
+import io.camunda.zeebe.protocol.record.ValueType;
+import io.camunda.zeebe.protocol.record.intent.Intent;
+import io.camunda.zeebe.util.buffer.BufferWriter;
+
+/**
+ * Represents a modifiable batch of record, which means we can add multiple Records to the batch.
+ * For further processing the user can iterate of the appended entries and retrieve the needed data.
+ */
+public interface ModifiableRecordBatch extends UnmodifiableRecordBatch {
+
+  /**
+   * Allows to add a new Record to the batch
+   *
+   * @param key the key of the record
+   * @param sourceIndex the position/index in the current batch which caused that entry; should be
```

Comment:
:+1: TIL - I wasn't sure what source index was. Is this a thing? Do we actually have entries in the same batch which cause others? How would that even happen? :thinking:

Reply:
I was not 100% sure whether it is really in use or just tests. But I kept it to not change to much for now. I will see after the abstraction thing, during the clean-up whether we can remove it. :)